src/components/graph/ExerciseGraphLayout.ts

```diff
@@ -0,0 +1,337 @@
+import { ExerciseGraph } from './ExerciseGraph'
+import { adjacencyMatrix, position, slug } from './graph-types'
+import { ExerciseGraphNode } from './ExerciseGraphNode'
+
+type options = {
+  width: number
+  padding: number
+  headerPadding: number
+  exerciseLayerHeight: number
+  conceptLayerHeight: number
+}
+
+export class ExerciseGraphLayout {
+  public graph: ExerciseGraph
+
+  // layout options
+  public width: number
+  public padding: number
+  public headerPadding: number
+  public exerciseLayerHeight: number
+  public conceptLayerHeight: number
+
+  public height: number
+
+  private adjacency: adjacencyMatrix
+
+  public exerciseLayers: Array<Array<ExerciseGraphNode>>
+  public conceptLayers: Array<Array<string>>
+
+  public exerciseLayerCount: number
+  public conceptLayerCount: number
+
+  public exercisePositions: Map<slug, position>
+  public conceptPositions: Map<string, position>
+
+  constructor(graph: ExerciseGraph, options: options) {
+    this.graph = graph
+
+    this.padding = options.padding
+    this.headerPadding = options.headerPadding
+    this.exerciseLayerHeight = options.exerciseLayerHeight
+    this.conceptLayerHeight = options.conceptLayerHeight
+    this.width = options.width
+
+    this.adjacency = makeAdjacencyMatrix(graph)
+
+    if (this.containsCycle()) {
+      throw new Error('🔥 Cannot continue. See error log.')
+    }
+
+    this.exerciseLayers = []
+    this.conceptLayers = []
+    this.computeNodeDepths()
+
+    this.exerciseLayerCount = this.exerciseLayers.length
+    this.conceptLayerCount = this.conceptLayers.filter(
+      (l) => l.length > 0
```

Comment:
_Why_ this works is unclear. You get it, I get it, but person X won't. Explain why there are layers with zero items? As a comment is fine.